owrx/websocket.py

```diff
@@ -175,60 +177,67 @@ def read_loop(self):
         def protected_read(num):
             data = self.handler.rfile.read(num)
             if data is None:
-                raise Drained()
-            if len(data) != num:
+                data = bytes()
+            while self.open and len(data) < num:
+                    (read, _, _) = select.select([self.interruptPipeRecv, self.handler.rfile], [], [], 15)
+                    if self.handler.rfile in read:
+                        data += self.handler.rfile.read(num - len(data))
+                    else:
+                        if len(data) == 0:
+                            raise InterruptOrTimeout()
+                        else:
+                            raise IncompleteRead()
+            if len(data) < num:
                 raise IncompleteRead()
             return data
 
         self.open = True
         while self.open:
-            (read, _, _) = select.select([self.interruptPipeRecv, self.handler.rfile], [], [], 15)
-            if self.handler.rfile in read:
-                available = True
+            try:
+                header = protected_read(2)
                 self.resetPing()
-                while self.open and available:
+                opcode = header[0] & 0x0F
+                length = header[1] & 0x7F
+                mask = (header[1] & 0x80) >> 7
+                if length == 126:
+                    header = protected_read(2)
+                    length = (header[0] << 8) + header[1]
+                if mask:
+                    masking_key = protected_read(4)
+                data = None
+                if length > 0:
```

Comment:
instead of implementing the check here, just pass `length=0` to `protected_read` and let it return early with an empty byte array. i'd say that's more elegant and removes some distracting conditions here.